torch/csrc/distributed/c10d/ProcessGroupUCC.cpp

```diff
@@ -0,0 +1,1575 @@
+    TORCH_CHECK(
+        false, "Size of Boolean type larger than 1 is not supported in UCC");
+  }
+  try {
+    return ucc_dtype_map.at(_tensor.scalar_type());
+  } catch (const std::out_of_range& e) {
+    TORCH_CHECK(false, "Not supported data type for UCC");
+  }
+}
+
+const std::map<ReduceOp, ucc_reduction_op_t> ucc_op_map = {
+    {ReduceOp::SUM, UCC_OP_SUM},
+    {ReduceOp::PRODUCT, UCC_OP_PROD},
+    {ReduceOp::MIN, UCC_OP_MIN},
+    {ReduceOp::MAX, UCC_OP_MAX},
+    {ReduceOp::BAND, UCC_OP_BAND},
+    {ReduceOp::BOR, UCC_OP_BOR},
+    {ReduceOp::BXOR, UCC_OP_BXOR},
+#if TORCH_VERSION_MAJOR > 1 || (TORCH_VERSION_MAJOR == 1 && TORCH_VERSION_MINOR >= 11)
+    {ReduceOp::AVG, UCC_OP_AVG},
+#endif
+};
+
+ucc_reduction_op_t to_ucc_reduceOp(
+    const ReduceOp _op,
+    const at::ScalarType _dt) {
+  if (_dt == at::kBool) {
+    if (_op == ReduceOp::SUM) {
+      // bitwise or
+      return UCC_OP_MAX;
+    } else if (_op == ReduceOp::PRODUCT) {
+      // bitwise and
+      return UCC_OP_MIN;
+    }
+#if TORCH_VERSION_MAJOR > 1 || \
+    (TORCH_VERSION_MAJOR == 1 && TORCH_VERSION_MINOR >= 11)
+    else if (_op == ReduceOp::AVG) {
+      TORCH_CHECK(
+          false, "Cannot use ReduceOp.AVG with boolean inputs");
+    }
+#endif
+  }
+
+  try {
+    return ucc_op_map.at(_op);
+  } catch (const std::out_of_range& e) {
+    TORCH_CHECK(
+        false, "Not supported ReduceOp for UCC");
+  }
+}
+
+struct torch_ucc_config_t {
+  std::once_flag flag;
+  std::array<bool, 32> blocking_wait;
+  bool enable_profiling;
+  bool use_future;
+  bool shared_comm;
+  bool use_allgatherv;
+} torch_ucc_config;
+
+std::map<std::string, std::string> torch_ucc_envs_map = {
+    {"TORCH_UCC_ALLGATHER_BLOCKING_WAIT", "1"},
+    {"TORCH_UCC_ALLREDUCE_BLOCKING_WAIT", "1"},
+    {"TORCH_UCC_ALLTOALL_BLOCKING_WAIT", "1"},
+    {"TORCH_UCC_BCAST_BLOCKING_WAIT", "1"},
+    {"TORCH_UCC_GATHER_BLOCKING_WAIT", "1"},
+    {"TORCH_UCC_REDUCE_SCATTER_BLOCKING_WAIT", "1"},
+    {"TORCH_UCC_SCATTER_BLOCKING_WAIT", "1"},
+    {"TORCH_UCC_USE_FUTURE", "1"},
+    {"TORCH_UCC_PROFILING_ENABLE", "0"},
+    {"TORCH_UCC_TLS", "nccl,ucp"},
+    {"TORCH_UCC_SHARED_COMM", "1"},
+    {"TORCH_UCC_USE_ALLGATHERV", "0"},
+};
+
+} // namespace
+
+void read_confg() {
+  // default configuration
+  torch_ucc_config.blocking_wait.fill(true);
+  torch_ucc_config.enable_profiling = false;
+  torch_ucc_config.use_future = true;
+  torch_ucc_config.shared_comm = false;
+  torch_ucc_config.use_allgatherv = false;
+
+  // read all torch_ucc env. variables and update the map
+  char* env;
+  for (auto& torch_ucc_env : torch_ucc_envs_map) {
+    env = std::getenv(torch_ucc_env.first.c_str());
+    if (env) {
+      torch_ucc_envs_map[torch_ucc_env.first] = std::string(env);
+    }
+  }
+  torch_ucc_config.blocking_wait[(std::uint8_t)OpType::ALLGATHER] =
+      std::stoi(torch_ucc_envs_map.at("TORCH_UCC_ALLGATHER_BLOCKING_WAIT"));
```

Comment:
Do we need a better name here? https://github.com/facebookresearch/torch_ucc/pull/77#discussion_r874102990

Reply:
I suggest to keep these names as is so that we have an easy cross-reference to the old torch-ucc repo. Then let's discuss about naming and send subsequent PRs to rename.